File: 3193-count-the-number-of-inversions/3193-count-the-number-of-inversions.java
class Solution {

    static int MOD = (int) 1e9 + 7;

    int dp[][];
    boolean vis[][];

    long solve(int n, HashMap<Integer, Integer> req, int index, int count) {

        if (index == 0) {
            if (count == 0) {
                return 1;
            }

            return 0;
        }

        if (req.containsKey(index)) {
            if (count != req.get(index)) {
                return 0;
            }
        }

        if (vis[index][count]) {
            return dp[index][count];
        }

        long ans = 0;

        for (int i = 0; i <= index; i++) {
            int contribution = i;

            if (contribution > count) {
                continue;
            }

            ans = ans + solve(n, req, index - 1, count - contribution);
            ans = ans % MOD;
        }

        vis[index][count] = true;
        dp[index][count] = (int) ans;

        return ans;
    }

    public int numberOfPermutations(int n, int[][] requirements) {

        HashMap<Integer, Integer> reqMap = new HashMap<>();

        int maxCount = 0;

        for (int i = 0; i < requirements.length; i++) {
            reqMap.put(requirements[i][0], requirements[i][1]);
            maxCount = Math.max(maxCount, requirements[i][1]);
        }

        if(reqMap.containsKey(0) && reqMap.get(0) > 0){
            return 0;
        }

        dp = new int[n + 1][maxCount + 1];
        vis = new boolean[n + 1][maxCount + 1];

        return (int) solve(n, reqMap, n - 1, reqMap.get(n - 1));
    }
}

File: 3193-count-the-number-of-inversions/3193-count-the-number-of-inversions.cpp
class Solution {
    const int M = 1e9+7;
public:
    int numberOfPermutations(int n, vector<vector<int>>& requirements) {
        vector<int> cnt(n, -1);
        sort(requirements.begin(), requirements.end());
        for(auto it: requirements){
            cnt[it[0]] = it[1];
        }

        vector<vector<int> > dp(n+1, vector<int>(405, 0));
        dp[0][0] = 1;
        if(cnt[0] != -1 && cnt[0] != 0) dp[0][0] = 0;

        for(int i = 1; i < n; i++){
            for(int j = 0; j <= 400; j++){
                for(int k = 0; k <= i; k++){
                    if(j+k <= 400) dp[i][j+k] = (dp[i][j+k]%M + dp[i-1][j]%M)%M;   
                }
                if(cnt[i] != -1 && cnt[i] != j) dp[i][j] = 0;
                dp[i][j] %= M;
            }
        }

        return dp[n-1][cnt[n-1]];
    }
};